mathjax3-ts/core/Tree/Factory.ts

```diff
@@ -0,0 +1,176 @@
+ */
+
+/*****************************************************************/
+/*
+ * The Factory node interfaces (one for the node instance, one for the node class)
+ */
+
+export interface FactoryNode {
+    readonly kind: string;
+}
+
+export interface FactoryNodeClass<N extends FactoryNode> {
+    new(factory: Factory<N, FactoryNodeClass<N>>, ...args: any[]): N;
+}
+
+/*****************************************************************/
+/*
+ * The Factory interface
+ */
+
+export interface Factory<N extends FactoryNode, C extends FactoryNodeClass<N>> {
+    /*
+     * @param {string} kind  The kind of node to create
+     * @return {N}  The newly created node of the given kind
+     */
+    create(kind: string): N;
+
+    /*
+     * Defines a class for a given node kind
+     *
+     * @param {string} kind  The kind whose class is being defined
+     * @param {C} nodeClass  The class for the given kind
+     */
+    setNodeClass(kind: string, nodeClass: C): void;
+    /*
+     * @param {string} kind  The kind of node whose class is to be returned
+     * @return {C}  The class object for the given kind
+     */
+    getNodeClass(kind: string): C;
+    /*
+     * @param {string} kind  The kind whose definition is to be deleted
+     */
+    deleteNodeClass(kind: string): void;
+
+    /*
+     * @param {N} node  The node to test if it is of a given kind
+     * @param {string} kind  The kind to test for
+     * @return {boolean}  True if the node is of the given kind, false otherwise
+     */
+    nodeIsKind(node: N, kind: string): boolean;
+
+    /*
+     * @return {string[]}  The names of all the available kinds of nodes
+     */
+    getKinds(): string[];
+}
+
+
+/*****************************************************************/
+/*
+ * The generic AbstractFactoryClass interface
+ *   (needed for access to defaultNodes via the constructor)
+ */
+
+interface AbstractFactoryClass<N extends FactoryNode, C extends FactoryNodeClass<N>> extends Function {
+    defaultNodes: {[kind: string]: C};
+}
+
+
+/*****************************************************************/
+/*
+ * The generic AbstractFactory class
+ */
+
+export abstract class AbstractFactory<N extends FactoryNode, C extends FactoryNodeClass<N>> implements Factory<N, C> {
+
+    /*
+     * The default collection of objects to use for the node map
+     */
+    public static defaultNodes = {};
+
+    /*
+     * The default kind
+     */
+    public defaultKind = 'unknown';
+
+    /*
+     * The map of node kinds to node classes
+     */
+    protected nodeMap: Map<string, C> = new Map();
+
+    /*
+     * An object containing functions for creating the various node kinds
+     */
+    protected node: {[kind: string]: (...args: any[]) => N} = {};
+
+    /*
+     * @override
+     */
+    constructor(nodes: {[kind: string]: C} = null) {
+        if (nodes === null) {
+            nodes = (this.constructor as AbstractFactoryClass<N, C>).defaultNodes;
```

Comment:
This is to gain access to the `defaultNodes` static property of whatever subclass is being instantiated).  I'm not sure what your comment is asking me to do.